File: mef_agri/app/web/src/msgs.js
/**
 * Class acting as namespace for messages which enable communication between 
 * frontend (PyQT) and map in WebEngine.
 * @class
 */
export class Messages {
    static keyMsgType = 'type';
    static keyMsgCont = 'content';

    ////////////////////////////////////////////////////////////////////////////
    // SEND-MESSAGES
    ////////////////////////////////////////////////////////////////////////////
    /**
     * Base class for messages which should be sent to the frontend
     */
    static SendMessage = class {
        #msg = {};

        constructor () {
            this.#msg[Messages.keyMsgType] = null;
            this.#msg[Messages.keyMsgCont] = {};
        }

        /**
         * There has to be a counterpart for message-content key in 
         * `mef_agri.app.gui.conn.msgs.Messages`, i.e. a class variable starting 
         * with `CONT_` in a nested class.
         * 
         * @param {string} key - key/identifier of a content field 
         * @param {any} val - value of a content field (should be JSON-serializable)
         */
        setCont(key, val) {
            this.#msg[Messages.keyMsgCont][key] = val;
        }

        /**
         * Setter for message-type.
         * 
         * There has to be a counterpart for the message-type in 
         * `mef_agri.app.gui.conn.msgs.Messages`, i.e. a class variable 
         * `MTYPE` of a nested class.
         * 
         * @param {string} val - type/identifier of message
         */
        set msgType(val) {
            this.#msg[Messages.keyMsgType] = val;
        }

        /**
         * @returns {object} - message (containing all information) which will be parsed with JSON and sent to the frontend
         */
        get message() {
            return this.#msg;
        }
    }

    /**
     * Class to send drawn field/polygons (coordinates and labels) to the 
     * frontend.
     * 
     * The crs/epsg-code is fixed to WGS84/4326.
     */
    static SendDrawnField = class extends Messages.SendMessage {
        constructor() {
            super();
            this.msgType = 'drawn_field';
            this.setCont('epsg', 4326);
        }

        /**
         * Setter for the field name.
         * 
         * @param {string} fname - name of the drawn field/polygon
         */
        set fieldName(fname) {
            this.setCont('fname', fname);
        }

        /**
         * Setter for the corner-coordinates of the drawn field/polygon
         * 
         * @param {Array} coords - array with corners/points (i.e. again arrays with two coordinates)
         */
        set cornerCoords(coords) {
            this.setCont('coords', coords);
        }
    };

    /**
     * Class to send log-messages to frontend
     */
    static SendLogMsg = class extends Messages.SendMessage {
        constructor() {
            super();
            this.msgType = 'logmsg';
        }
        
        /**
         * setter
         * @param {string} msg - log-message which should be sent to the frontend
         */
        set logMsg(msg) {
            this.setCont('logmsg', msg);
        }
    }

    ////////////////////////////////////////////////////////////////////////////
    // INCOMING-MESSAGES
    ////////////////////////////////////////////////////////////////////////////
    /**
     * Base class for messages which come from the frontend.
     */
    static GotMessage = class {
        #msgCont;
        constructor(msg) {
            this.#msgCont = msg[Messages.keyMsgCont];
        }

        /**
         * @param {string} key - key specifying the content field for which a value should be returned
         * @returns {any} - value which belongs to the content-field
         */
        getCont(key) {
            return this.#msgCont[key];
        }
    }

    /**
     * Class which handles incoming messages containing information about 
     * available fields.
     */
    static GeoFieldInfo = class extends Messages.GotMessage {
        static msgType = 'field_defs';

        /**
         * @returns {Array} - array with field names being strings
         */
        get fieldNames() {
            return this.getCont('fnames');
        }

        /**
         * @returns {Array} - array containing polygon-defining corner-coordinate-lists - defined in WGS84
         */
        get cornerCoords() {
            return this.getCont('coords');
        }
    }
}

File: mef_agri/app/gui/conn/msgs.py
import json


class MsgBaseClass(object):
    MTYPE = None

    def __init__(self, msgcont:dict=None):
        """
        :param msgcont: message with appropriate content (if available), defaults to None
        :type msgcont: dict, optional
        """
        if msgcont is None:
            msgcont = {}
        for aname, aval in self.__class__.__dict__.items():
            if aname.startswith('CONT_'):
                if aval in msgcont.keys():
                    val = msgcont[aval]
                else:
                    val = None
                setattr(self, '_' + aval, val)

    @property
    def message(self) -> str:
        """
        :return: message which should be passed to websocket
        :rtype: str
        """
        msg = {Messages.KEY_MTYPE: self.MTYPE}
        cont = {}
        for aname, aval in self.__class__.__dict__.items():
            if aname.startswith('CONT_'):
                cont[aval] = getattr(self, '_' + aval)
        msg[Messages.KEY_CONT] = cont
        return json.dumps(msg)


class Messages:
    KEY_MTYPE = 'type'
    KEY_CONT = 'content'

    class GotDrawnField(MsgBaseClass):
        MTYPE = 'drawn_field'
        CONT_COORDS = 'coords'
        CONT_EPSG = 'epsg'
        CONT_FNAME = 'fname'

        @property
        def points(self) -> list:
            """
            :return: coordinates of points representing the polygon
            :rtype: list
            """
            return getattr(self, '_' + self.CONT_COORDS)
        
        @property
        def epsg(self) -> int:
            """
            :return: epsg-code of crs in which :func:`points` are defined
            :rtype: int
            """
            return getattr(self, '_' + self.CONT_EPSG)
        
        @property
        def field_name(self) -> str:
            """
            :return: name of the field
            :rtype: str
            """
            return getattr(self, '_' + self.CONT_FNAME)
        
    class GotDeleteField(MsgBaseClass):
        MTYPE = 'delete_field'
        CONT_FNAME = 'fname'

        @property
        def field_name(self) -> str:
            """
            :return: name of the field which should be deleted
            :rtype: str
            """
            return getattr(self, '_' + self.CONT_FNAME)

    class GotLogMsg(MsgBaseClass):
        MTYPE = 'logmsg'
        CONT_LOGMSG = 'logmsg'

        @property
        def log_message(self) -> str:
            return getattr(self, '_' + self.CONT_LOGMSG)

    class SendFields(MsgBaseClass):
        MTYPE = 'field_defs'
        CONT_FNAMES = 'fnames'
        CONT_COORDS = 'coords'

        @property
        def field_names(self) -> list[str]:
            """
            :return: (unique) name of the field
            :rtype: str
            """
            return getattr(self, '_' + self.CONT_FNAMES)
        
        @field_names.setter
        def field_names(self, value):
            setattr(self, '_' + self.CONT_FNAMES, value)

        @property
        def coordinates(self) -> list:
            """
            :return: list of polygon-defining point-coordinate-lists - coordinates have to be defined in web-mercator (epsg:3857)
            :rtype: list[list[list[float, float]]]
            """
            return getattr(self, '_' + self.CONT_COORDS)
        
        @coordinates.setter
        def coordinates(self, value):
            setattr(self, '_' + self.CONT_COORDS, value)

    class SendActiveTab(MsgBaseClass):
        MTYPE = 'active_tab'
        CONT_TABNAME = 'tabname'

        @property
        def tab_name(self) -> str:
            """
            :return: name of the currently active tab
            :rtype: str
            """
            return getattr(self, '_' + self.CONT_TABNAME)
        
        @tab_name.setter
        def tab_name(self, value):
            setattr(self, '_' + self.CONT_TABNAME, value)
